protocols/request-response/tests/error_reporting.rs

```diff
@@ -0,0 +1,727 @@
+                        .send_response(channel, Action::FailOnReadResponse)
+                        .unwrap();
+                }
+                Ok(request_response::Event::ResponseSent {
+                    peer, request_id, ..
+                }) => {
+                    assert_eq!(peer, peer2_id);
+                    assert_eq!(req_id, Some(request_id));
+                }
+                Ok(ev) => panic!("Peer1: Unexpected event: {ev:?}"),
+                Err(..) => {}
+            }
+        }
+    };
+
+    // Client
+    //
+    // Expects OutboundFailure::Io failure with `FailOnReadResponse` error
+    let swarm2_task = async move {
+        let req_id = swarm2
+            .behaviour_mut()
+            .send_request(&peer1_id, Action::FailOnReadResponse);
+
+        loop {
+            match swarm2.select_next_some().await.try_into_behaviour_event() {
+                Ok(request_response::Event::OutboundFailure {
+                    peer,
+                    request_id,
+                    error,
+                }) => {
+                    assert_eq!(peer, peer1_id);
+                    assert_eq!(request_id, req_id);
+
+                    let error = match error {
+                        OutboundFailure::Io(e) => e,
+                        e => panic!("Peer2: Unexpected error {e:?}"),
+                    };
+
+                    assert_eq!(error.kind(), io::ErrorKind::Other);
+                    assert_eq!(
+                        error.into_inner().unwrap().to_string(),
+                        "FailOnReadResponse"
+                    );
+                    break;
+                }
+                Ok(ev) => panic!("Peer2: Unexpected event: {ev:?}"),
+                Err(..) => {}
+            }
+        }
+    };
+
+    spawn(swarm1_task);
+    timeout(Duration::from_millis(100), swarm2_task)
+        .await
+        .expect("timed out on waiting FailOnReadResponse");
+
+    assert!(!panic_check_rx.is_closed(), "swarm1_task panicked");
+}
+
+#[async_std::test]
+async fn report_outbound_failure_on_write_request() {
+    let _ = env_logger::try_init();
+
+    let protocols = iter::once((StreamProtocol::new("/test/1"), ProtocolSupport::Full));
+    let cfg = request_response::Config::default();
+
+    let mut swarm1 = Swarm::new_ephemeral(|_| {
+        request_response::Behaviour::<TestCodec>::new(protocols.clone(), cfg.clone())
+    });
+    let peer1_id = *swarm1.local_peer_id();
+
+    let mut swarm2 =
+        Swarm::new_ephemeral(|_| request_response::Behaviour::<TestCodec>::new(protocols, cfg));
+
+    swarm1.listen().await;
+    swarm2.connect(&mut swarm1).await;
+
+    // On panic `panic_check_rx` will be closed
+    let (panic_check_tx, panic_check_rx) = channel::bounded::<()>(1);
+
+    // Server
+    //
+    // Expects no events because `Event::Request` is produced after `read_request`.
+    let swarm1_task = async move {
+        let _panic_check_tx = panic_check_tx;
```

Comment:
Instead of the panic check, couldn't we just return an error from the future and then check what it finished with?

Reply:
I use panic checks used for the tasks that never finished. Do you mean to not spawning a task at all? I thought for the swarm to progress, both needed to be polled at the same time.

Reply:
Yeah you'll need to poll them. I think what I'd do is spawn the task but grab the `JoinHandle` you get returned. You can check on that whether the task is finished or not so you don't need the separate channel I think.

Reply:
I did it with `futures::future::select`